jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/ObtainBaseImageLayerStep.java

```diff
@@ -36,14 +38,50 @@
 import javax.annotation.Nullable;
 
 /** Pulls and caches a single base image layer. */
-class PullAndCacheBaseImageLayerStep implements Callable<CachedLayerAndName> {
+class ObtainBaseImageLayerStep implements Callable<PreparedLayer> {
 
   private static final String DESCRIPTION = "Pulling base image layer %s";
 
-  static ImmutableList<PullAndCacheBaseImageLayerStep> makeList(
+  @FunctionalInterface
+  private interface BlobExistenceChecker {
+
+    Optional<Boolean> exists(DescriptorDigest digest) throws IOException, RegistryException;
+  }
+
+  static ImmutableList<ObtainBaseImageLayerStep> makeListForForcedDownload(
       BuildConfiguration buildConfiguration,
       ProgressEventDispatcher.Factory progressEventDispatcherFactory,
       ImageAndAuthorization baseImageAndAuth) {
+    BlobExistenceChecker noOpBlobChecker = ignored -> Optional.empty();
+    return makeList(
+        buildConfiguration, progressEventDispatcherFactory, baseImageAndAuth, noOpBlobChecker);
+  }
+
+  static ImmutableList<ObtainBaseImageLayerStep> makeListForSelectiveDownload(
+      BuildConfiguration buildConfiguration,
+      ProgressEventDispatcher.Factory progressEventDispatcherFactory,
+      ImageAndAuthorization baseImageAndAuth,
+      Authorization pushAuthorization) {
+    Verify.verify(!buildConfiguration.isOffline());
+
+    RegistryClient targetRegistryClient =
+        buildConfiguration
+            .newTargetImageRegistryClientFactory()
+            .setAuthorization(pushAuthorization)
+            .newRegistryClient();
+    // TODO: also check if cross-repo blob mount is possible.
```

Comment:
I wonder if this is as easy as adding `buildConfiguration.getBaseImageRegistry().equals(buildConfiguration.getTargetImageRegistry())` on in the existing-checker below?

Reply:
I thought it involves more than that, like checking
```
JibSystemProperties.useCrossRepositoryBlobMounts()
    && canAttemptBlobMount(authorization, sourceRepository))
```
to ensure that we will attempt the blob mount. And I think it is that we can *attempt* blob mount, and it might be possible that a server may not allow/support the mount by returning `202 Accepted` instead of `201 Created`, in which case `RegistryClient.pushBlob()` will fall back to do the usual pushing. So I think this is actually complicated.

But I think we are already pretty good without blob mount. Most of the time, the base image will be in the target repository. The blob mount can be a further optimization, but I think it does not give a huge value.